### PR TITLE
Adding support for symlinks

### LINKED ISSUE #6094: ADD to support symlinks in the <src> argument
(from #1676)

In a Dockerfile, when `ADD`ing files, symlinks are not followed. Okay, it's insecure to follow symlinks, moreover -- some might want to put a symlink as a symlink!

What I suggest is that `ADD` follows intermediate symlinks: e.g. 

```
ADD a/b/c /path/to/
```

should work fine, if `a` or `b` or `c` is a symlink, since it needs not to be transferred

### PR DESCRIPTION
**Pull request by brahmaroutu**
Fixes #6094

Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

File: integration-cli/docker_cli_build_test.go [hunk 1]
```diff
@@ -727,6 +727,117 @@ RUN [ $(ls -l /exists/exists_file | awk '{print $3":"$4}') = 'dockerio:dockerio'
...
 }
 
+func TestBuildAddLinkToDir(t *testing.T) {
+	name := "testaddlinktoexistdir"
+	defer deleteImages(name)
+	ctx, err := fakeContext(`FROM busybox
+ADD exists/test_file test_file
+ADD link2dir lexists
+RUN ls -L test_file lexists lexists/test_file  
+[ "$(readlink /lexists)" == "test_file" ]`,
```
Comment (LK4D4) [inline integration-cli/docker_cli_build_test.go]: This is not really dockerfile instruction.

**Comment by tianon** [comment]
I'm -1 on this -- I think this would be more "magic" in `COPY`/`ADD` where just copying the symlink as a symlink would be the expected behavior.

**Comment by tiborvass** [comment]
* Review session with @jfrazelle @LK4D4 @tianon *

@brahmaroutu Thanks for your patience. We are okay with the functionality, provided that we add more tests.

Basically, we need to make sure that a symlink is dereferenced only if it is followed by a `/`.
In the example `ADD a/b/c /path/to/`, if `a`, `b`, and `c` are all symlinks, then `a` and `b` are followed, but not `c`, creating `/path/to/c` that is the same symlink as `c`.

```
$ cd /tmp/test
$ ls
$ mkdir A B
$ ln -s A a
$ ln -s /B A/b
$ ln -s /A/C B/c
$ echo 'from busybox
copy a/b/c /path/to/
run ls -l /path/to' > Dockerfile
$ docker build -t test .   # this fails with your PR but should work and show "c -> /A/C"
Sending build context to Docker daemon 4.608 kB
Sending build context to Docker daemon
Step 0 : FROM busybox
 ---> 4986bf8c1536
Step 1 : COPY a/b/c /path/to/
INFO[0000] a/b/c: no such file or directory
```

If you are not available, we can carry your PR preserving your commit; let us know.

**Comment by brahmaroutu** [comment]
@tiborvass I can add the test but will take few hours to get to it.

**Comment by brahmaroutu** [comment]
@tiborvass your example does not seem to work manually either, I am confused. I cannot do "ls -l a/b/c"

**Comment by tiborvass** [comment]
@brahmaroutu Ok, so it seems that we cannot follow symlinks out of our context, and we shouldn't.

So my previous example that was using absolute symlinks should always fail.

I'm rewriting the example that I think should work, but that still does not work in your PR:
```
$ mkdir A B
$ echo foo > B/C
$ ln -sf A a; ln -sf ../B A/b; ln -sf C B/c
$ echo 'from busybox
copy a/b/c /path/to/
run ls -l /path/to' > Dockerfile
$ docker build --no-cache .        # should show "c -> C" and not just "c" in the ls.
Sending build context to Docker daemon 5.632 kB
Sending build context to Docker daemon
Step 0 : FROM busybox
 ---> 4986bf8c1536
Step 1 : COPY a/b/c /path/to/
 ---> 2a2494cb854a
Removing intermediate container 7d8dccf59489
Step 2 : RUN ls -l /path/to
 ---> Running in 6949978707ac
total 4
-rw-r--r--    1 root     root             4 Jan 14 00:50 c
 ---> 85860c691789
Removing intermediate container 6949978707ac
Successfully built 85860c691789
```

Hope this helps, let me know if you have any questions

**Comment by brahmaroutu** [comment]
I did the relative links test and also add functionality to have support multiple links as you stated above. Please review. I may have to add more test cases for copy,etc.

**Comment by tiborvass** [comment]
@brahmaroutu with the same example, now it shows `C` where previously it was outputting `c`, but what I want to see, is `c -> C` (sorry for the rhymes).

**Comment by brahmaroutu** [comment]
@tiborvass If the top level link resolves to a file it will copy the file removing the link. This is the current behavior. If the link is inside a normal directory then it preserves the link.

**Comment by tiborvass** [comment]
@brahmaroutu Ok, I confirm what you said.

@tianon if c is a symlink in a/b/c then we do resolve it currently, so what we're suggesting would be a breaking change essentially :( So what do you think we should do ?

**Comment by brahmaroutu** [comment]
Wow, thats funny. Does that mean we have a go?

**Comment by tiborvass** [comment]
/cc @duglin @erikh please weigh in

@brahmaroutu If there's `a/b/c` where only `c` is a symlink, and you copy `a/b`, then `c` is preserved inside the container and not resolved. So with @tianon we're thinking of doing the breaking change, since the symlink behavior is neither documented nor tested and seems to be an incorrect behavior.
This means in practice that when copied, we do want `a/b/c` to be preserved as a symlink. The rule being: *if it is followed by a slash, then resolve the symlink*.

Sorry if it's not super clear, I'm a little tired. Let me know if you need clarifications.

**Comment by SvenDowideit** [comment]
This cannot be considered for merging until there is some very clear documentation explaining to users how and when it will and won't work. and in the process, you'd be showing eg's of when its useful.

I can make some pretty convoluted guesses as to how i would use it, but we don't want users guessing.

**Comment by brahmaroutu** [comment]
@tiborvass @tianon As I said before the current behavior works differently for links to file vs directory and do not preserve symlinks, I am not attempting to preserve symlinks and that  probably requires some serious design considerations.

Todays functionality is as follows:  (links to dir drink and link to a file)
drwxrwxr-x 2 vagrant vagrant 4096 Jan 13 23:11 A  //directory here has files in it 
lrwxrwxrwx 1 vagrant vagrant    1 Jan 15 03:40 dlink -> A
-rw-rw-r-- 1 vagrant vagrant   10 Jan 15 03:39 file1
lrwxrwxrwx 1 vagrant vagrant    5 Jan 15 03:40 flink -> file1

When I add    "ADD dlink DL"  && "ADD flink FL"  to docker file I get the following  
drwxr-xr-x   2 root root 4096 Jan 15 03:40 DL
-rw-rw-r--   1 root root   10 Jan 15 03:39 FL
root@93963bed52e9:/# more FL  //copied content of the file
test file
root@93963bed52e9:/# ls DL  //created a empty dir for what?
root@93963bed52e9:/#      

I am only fixing the issue that symlink to a directory does not copy entire directory but creates a empty directory but for a link to file it actually created the physical file with content.

I terms of the documentation, if you let me know where the current behavior is described I can add content for the directory part.

Please let me know if you want to do it differently, we can close this PR.

**Comment by SvenDowideit** [comment]
@tiborvass @tianon it sounds to me like the existing functionality is quite odd, and also undocumented.

@crosbymichael @vieux do we have any memory of why it works the way it does, and why its not documented?

@brahmaroutu I'm sorry, I misconstrued your description: "Fixes #6094" as supporting symlinks in ADD

it looks to me like the existing function is undocumented, and that you have a grasp on what it does, so I hope you can add that knowledge to the builder.md ADD section.

I don't entirely understand what the problems here are - I'm wondering if `ADD symlink /somewhere` shoudln't just be naively copied into the container without evaluation - so that the symlink is pointing to that location inside the container fs (and can be dangling, pointing to nothing)

**Comment by tiborvass** [comment]
@brahmaroutu OK, now I understand better. I'm sorry it took us time, symlinks are always tricky, and mixed with this undocumented weird behavior didn't help.

We have only two options:
* We accept this PR, that will make the behavior with a directory symlink more consistent with the file symlink. I agree that the directory behavior as mentioned in https://github.com/docker/docker/pull/8420#issuecomment-70036558 is something to be fixed.
* We close this PR and decide that we would like to fix the symlink behavior once and for all, even if it means we're doing a breaking change (of an undocumented, weird behavior). That is, copying a symlink does not follow the symlink, simply copies the symlink (duh). Only follow a symlink if there is a `/` after it.

Ping @tianon @jfrazelle @icecrime @unclejack @erikh @LK4D4 @duglin please weigh into this design decision.

**Comment by tianon** [comment]
I'm +1 for option 2.

**Comment by brahmaroutu** [comment]
I am fine with either of the options. But with option 2 let us say we have "ADD /a/b/c A" where c is a link to /A/B/C/D then does it mean that we want to copy a/b/c and A/B/C/D so that we preserve the symlinks? If so that will also change the behavior for symlinks to files. We end up copying more than the target into the container?
@SvenDowideit I understand that we need to document this behavior no matter which path we pick, I will work on it as we clarify the approach.

**Comment by SvenDowideit** [comment]
@brahmaroutu awesome, thank you very much!

**Comment by tiborvass** [comment]
@brahmaroutu sorry for replying only now.

If I take your example, `ADD /a/b/c A` would create a file named A that is a symlink, instead of the target. So the answer is no, it would not copy both, only the symlink. That's why it would be a breaking change (that we could think of as a bugfix).

Another option is to keep the behavior on ADD and differentiate it with COPY. /cc @tianon 

@brahmaroutu I suggest we close this PR and open a new one, so that the discussion is not bloated.

**Comment by tiborvass** [comment]
I'm closing this for now, if there are any concerns feel free to comment below

**Pull request closed without merging**